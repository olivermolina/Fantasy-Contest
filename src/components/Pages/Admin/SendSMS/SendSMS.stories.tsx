import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';

import SendSMS from './SendSMS';

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  component: SendSMS,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as ComponentMeta<typeof SendSMS>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof SendSMS> = (args) => (
  <SendSMS {...args} />
);

export const ManageSendSMS = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
ManageSendSMS.args = {
  onSubmit: () => console.log('submit'),
};
