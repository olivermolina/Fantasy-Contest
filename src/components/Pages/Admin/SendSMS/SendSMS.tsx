import React, { useEffect } from 'react';
import { useTheme } from '@mui/material/styles';
import { Box, Button, FormHelperText, TextField } from '@mui/material';
import { useForm } from 'react-hook-form';
import useMediaQuery from '@mui/material/useMediaQuery';
import * as z from 'zod';
import OutlinedInput from '@mui/material/OutlinedInput';
import InputLabel from '@mui/material/InputLabel';
import MenuItem from '@mui/material/MenuItem';
import FormControl from '@mui/material/FormControl';
import Chip from '@mui/material/Chip';
import Select, { SelectChangeEvent } from '@mui/material/Select';
import Checkbox from '@mui/material/Checkbox';
import ListItemText from '@mui/material/ListItemText';
import DeleteIcon from '@mui/icons-material/Delete';
import { zodResolver } from '@hookform/resolvers/zod';
import { SendSMSInputValidationSchema } from '~/schemas/SendSMSInputValidationSchema';

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;
const MenuProps = {
  PaperProps: {
    style: {
      maxHeight: ITEM_HEIGHT * 4.5 + ITEM_PADDING_TOP,
      width: 250,
    },
  },
};

/**
 * Defines the form fields for SendSMS form.
 */
export type SendSMSInput = z.infer<typeof SendSMSInputValidationSchema>;

export interface SMSUser {
  id: string;
  email: string;
  username: string | null;
  phone: number | null;
}

interface SendSMSProps {
  /**
   * Submit form function
   */
  onSubmit: (inputs: SendSMSInput) => void;
  users: SMSUser[];
}

export const SELECT_ALL: SMSUser = {
  id: 'SELECT_ALL',
  email: '',
  username: 'All Users',
  phone: null,
};

const SendSMS = (props: SendSMSProps) => {
  const { onSubmit, users } = props;
  const [userSelect, setUserSelect] = React.useState<string[]>([]);
  const handleRemove = (userId: string) => {
    setUserSelect((prev) =>
      prev.filter((item) => item !== userId && item !== SELECT_ALL.id),
    );
  };

  const handleChange = (event: SelectChangeEvent<typeof userSelect>) => {
    const {
      target: { value },
    } = event;

    // Select All
    if (
      value.indexOf(SELECT_ALL.id) > -1 &&
      userSelect.indexOf(SELECT_ALL.id) === -1
    ) {
      setUserSelect([SELECT_ALL.id]);
      return;
    }

    // Deselect All
    if (
      userSelect.indexOf(SELECT_ALL.id) > -1 &&
      value.indexOf(SELECT_ALL.id) === -1
    ) {
      setUserSelect([]);
      return;
    }

    const newValue =
      // On autofill we get a stringified value.
      typeof value === 'string' ? value.split(',') : value;

    setUserSelect(
      newValue.length === users.length + 1
        ? newValue
        : newValue.filter((name) => name !== SELECT_ALL.id),
    );
  };

  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.down('md'));

  const {
    register,
    handleSubmit,
    formState: { errors },
    setValue,
  } = useForm<SendSMSInput>({
    resolver: zodResolver(SendSMSInputValidationSchema),
  });

  useEffect(() => {
    setValue('userIds', userSelect as [string, ...string[]]);
  }, [userSelect]);

  return (
    <div className={'flex flex-col gap-2 p-2'}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className={`flex flex-col items-start gap-2`}>
          <FormControl fullWidth error={!!errors?.userIds}>
            <InputLabel id="select-users-label" shrink={true}>
              Users
            </InputLabel>
            <Select
              labelId="select-users-label"
              id="select-users"
              multiple
              value={userSelect}
              onChange={handleChange}
              input={<OutlinedInput id="select-multiple-chip" label="Chip" />}
              renderValue={(selected) => (
                <Box sx={{ display: 'flex', flexWrap: 'wrap', gap: 0.5 }}>
                  {selected.map((value) => (
                    <Chip
                      key={value}
                      label={
                        [...users, SELECT_ALL].find((user) => user.id === value)
                          ?.username
                      }
                      onDelete={() => {
                        handleRemove(value);
                      }}
                      onMouseDown={(event) => {
                        event.stopPropagation();
                      }}
                      deleteIcon={<DeleteIcon />}
                    />
                  ))}
                </Box>
              )}
              MenuProps={MenuProps}
              notched={true}
            >
              <MenuItem value={SELECT_ALL.id}>
                <Checkbox checked={userSelect.indexOf(SELECT_ALL.id) > -1} />
                <ListItemText
                  primary={SELECT_ALL.username}
                  className={'italic text-gray-400'}
                />
              </MenuItem>

              {users.map((user) => (
                <MenuItem key={user.id} value={user.id}>
                  <Checkbox checked={userSelect.indexOf(user.id) > -1} />
                  <ListItemText
                    primary={
                      <>
                        {user.username}{' '}
                        <em className={'italic text-gray-400'}>
                          {user.phone?.toString()}
                        </em>
                      </>
                    }
                  />
                </MenuItem>
              ))}
            </Select>
            <FormHelperText>{errors?.userIds?.message}</FormHelperText>
          </FormControl>
          <TextField
            id="outlined-basic"
            label="SMS Body"
            variant="outlined"
            type={'number'}
            fullWidth
            {...register('textMessage')}
            error={!!errors?.textMessage}
            helperText={errors?.textMessage?.message}
            size={'small'}
            multiline
            rows={10}
          />
          <Button
            variant="contained"
            type="submit"
            fullWidth={matches}
            name={'submit'}
          >
            Send
          </Button>
        </div>
      </form>
    </div>
  );
};

export default SendSMS;
