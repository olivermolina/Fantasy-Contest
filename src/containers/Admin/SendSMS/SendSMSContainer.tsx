import { trpc } from '~/utils/trpc';
import React, { useMemo } from 'react';
import { SubmitHandler } from 'react-hook-form';
import { toast } from 'react-toastify';
import { TRPCClientError } from '@trpc/client';
import BackdropLoading from '~/components/BackdropLoading';
import SendSMS, {
  SELECT_ALL,
  SendSMSInput,
  SMSUser,
} from '~/components/Pages/Admin/SendSMS/SendSMS';

const SendSMSContainer = () => {
  const mutation = trpc.admin.createSms.useMutation();

  const { data, isLoading } = trpc.user.users.useQuery();

  const onSubmit: SubmitHandler<SendSMSInput> = async (inputs) => {
    try {
      const sendToAll = inputs.userIds.includes(SELECT_ALL.id);
      await mutation.mutateAsync({
        ...inputs,
        userIds: sendToAll
          ? (data?.map((user) => user.id) as [string, ...string[]]) || []
          : inputs.userIds,
      });
      toast.success(`SMS sent successfully`);
    } catch (error) {
      const e = error as TRPCClientError<any>;
      toast.error(
        e?.message || `Oops! Something went wrong. Please try again later.`,
      );
    }
  };

  const usersWithPhone = useMemo(() => {
    return (data?.filter((user) => user.phone) as SMSUser[]) || [];
  }, [data]);

  return (
    <>
      <BackdropLoading open={mutation.isLoading || isLoading} />
      <SendSMS onSubmit={onSubmit} users={usersWithPhone} />
    </>
  );
};

export default SendSMSContainer;
