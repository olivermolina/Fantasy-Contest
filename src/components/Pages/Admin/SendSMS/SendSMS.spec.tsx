import React from 'react';
import { render } from '@testing-library/react';
import SendSMS from './SendSMS';

describe('SendSMS', () => {
  it('should render', () => {
    const { container } = render(
      <SendSMS onSubmit={() => console.log('submit')} />,
    );
    expect(container).toMatchSnapshot();
  });
});
