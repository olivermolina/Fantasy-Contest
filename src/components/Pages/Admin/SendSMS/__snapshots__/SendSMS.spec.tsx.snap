// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendSMS should render 1`] = `
<div>
  <div
    class="flex flex-col gap-2 p-2"
  >
    <form>
      <div
        class="flex flex-col items-start gap-2"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="select-users-label"
          >
            Users
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="select-users-label select-users"
              class="MuiSelect-select MuiSelect-outlined MuiSelect-multiple MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="select-users"
              role="button"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              id="select-multiple-chip"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Chip
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
          />
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            SMS Body
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-z1o969-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              name="textMessage"
              rows="10"
              style="height: 0px; overflow: hidden;"
            />
            <textarea
              aria-hidden="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  SMS Body
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          name="submit"
          tabindex="0"
          type="submit"
        >
          Send
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</div>
`;
